internal/handlers/common/sort.go

```diff
@@ -14,10 +14,159 @@
 
 package common
 
-import "github.com/FerretDB/FerretDB/internal/types"
+import (
+	"sort"
+	"strings"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// sortType represents sort type
+type sortType int
+
+const (
+	ascending sortType = iota
+	descending
+)
 
 // SortDocuments sorts given documents in place according to the given sorting conditions.
 func SortDocuments(docs []*types.Document, sort *types.Document) error {
-	// TODO
+	if sort.Len() == 0 {
+		return nil
+	}
+
+	if sort.Len() > 32 {
+		return lazyerrors.Errorf("maximum sort keys exceeded: %v", sort.Len())
+	}
+
+	sortFuncs := make([]sortFunc, len(sort.Keys()))
+	for i, sortKey := range sort.Keys() {
+		sortField, err := sort.Get(sortKey)
+		if err != nil {
+			return err
+		}
+		sortType, err := getSortType(sortField)
+		if err != nil {
+			return err
+		}
+
+		sortFuncs[i] = func(a, b *types.Document) bool {
+			sortKey := sortKey
+			sortType := sortType
+			aField, err := a.Get(sortKey)
+			if err != nil {
+				return false
+			}
+
+			bField, err := b.Get(sortKey)
+			if err != nil {
+				return false
+			}
+
+			switch aField.(type) {
+			case string:
+				aField, err := AssertType[string](aField)
+				if err != nil {
+					return false
+				}
+				bField, err := AssertType[string](bField)
+				if err != nil {
+					return false
+				}
+
+				return strings.Compare(aField, bField) == -1
+			default:
+				result := compareScalars(aField, bField)
+				return matchSortResult(sortType, result)
+			}
+		}
+	}
+
+	sorter := &docsSorter{docs: docs, sorts: sortFuncs}
+	sorter.Sort(docs)
+
 	return nil
 }
+
+type sortFunc func(a, b *types.Document) bool
+
+type docsSorter struct {
+	docs  []*types.Document
+	sorts []sortFunc
+}
+
+func (ds *docsSorter) Sort(docs []*types.Document) {
+	ds.docs = docs
+	sort.Sort(ds)
+}
+
+func (ds *docsSorter) Len() int {
+	return len(ds.docs)
+}
+
+func (ds *docsSorter) Swap(i, j int) {
+	ds.docs[i], ds.docs[j] = ds.docs[j], ds.docs[i]
+}
+
+func (ds *docsSorter) Less(i, j int) bool {
+	p, q := ds.docs[i], ds.docs[j]
+	// Try all but the last comparison.
+	var k int
+	for k = 0; k < len(ds.sorts)-1; k++ {
+		sortFunc := ds.sorts[k]
+		switch {
+		case sortFunc(p, q):
+			// p < q, so we have a decision.
+			return true
+		case sortFunc(q, p):
+			// p > q, so we have a decision.
+			return false
+		}
+		// p == q; try the next comparison.
+	}
+	// All comparisons to here said "equal", so just return whatever
+	// the final comparison reports.
+	return ds.sorts[k](p, q)
+}
+
+func getSortType(value any) (sortType, error) {
+	switch value.(type) {
+	case int32:
```

Comment:
There might be the same problem as #379: it is possible that MongoDB can accept any whole number there, including int64 and float64, similar to https://github.com/FerretDB/FerretDB/blob/b0804b43a0fae38be0b00e503733981af8ad5528/internal/handlers/common/filter.go#L304-L317

But we should check that with dance tests first

Reply:
Checked with dance tests. Updated getSortType.